File: Problem-2_Basic_calculator_ii.py
# APPROACH  
# Time Complexity : O(s), s: len(s)
# Space Complexity : O(s)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : None
#
#
# Your code here along with comments explaining your approach
# 1. if char is int, save it to curr_num (maintain 42 and not 4 and 2)
# 2. if char is sign -> check if the last sign is / or * -> pop from stack and apply the op on it with the curr_num
#                    -> if last sign is + or -, push the curr_num to stack along with it's sign
# 3. if char is space -> IGNORE
# 4. if the last char, push to stack along with the sign (+ or -)
# 5. at end, sum all elements of stack -> result

class Solution:
    def calculate(self, s: str) -> int:
        
        if s is None:
            return None
        
        stack, curr_num, last_sign = [], 0, '+'
        
        for ind in range(len(s)):
            
            if ord(s[ind]) >= ord('0') and ord(s[ind]) <= ord('9'):
                curr_num = curr_num * 10 + (ord(s[ind]) - ord('0'))
        
            if ((ord(s[ind]) < ord('0') or ord(s[ind]) > ord('9')) and s[ind] != ' ') or ind == len(s) - 1:
                
                if last_sign == '/':
                    popped, flag = stack.pop(), False
                    if popped < 0: flag = True
                    res = abs(popped) // curr_num
                    if flag: stack.append(-1 * res)
                    else: stack.append(res)
                        
                elif last_sign == '*': stack.append(stack.pop() * curr_num)
                elif last_sign == '+': stack.append(curr_num)
                elif last_sign == '-': stack.append(-1 * curr_num)
                    
                curr_num = 0
                last_sign = s[ind]
                
        result = 0
        while stack:
            result += stack.pop()
            
        return result

File: Problem2_BasicCalculatorii.java
// Time Complexity : o(n)
// Space Complexity : 0(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

class Solution {
    public int calculate(String s) {
        
        if(s==null||s.length()==0)
            return 0;
        
        Stack<Integer> stack=new Stack<>();
        
        s=s.trim()+'+';
        s=s.replace(" ","");
        
        int result=0;
        
        int num=0;
        char lastSign='+';
        
        for(int i=0;i<s.length();i++)
        {
            char ch=s.charAt(i);
            
            //if its digit
            if(Character.isDigit(ch))
                num=num*10+ch-'0';
            // not a digit 
            else
            {
                if(lastSign=='+')
                {
                    stack.push(num);
                }
                else if(lastSign=='-')
                {
                    stack.push(-num);
                }
                else if(lastSign=='*')
                {
                    stack.push(stack.pop()*num);
                }
                else
                {
                    stack.push(stack.pop()/num);
                }
                num=0;
                lastSign=ch;
            }
        }
        
        while(!stack.isEmpty())
        {
            result+=stack.pop();
        }
        return result;
    }
}